service/gitea/gitea.go

```diff
@@ -100,25 +99,242 @@ func (g *PullRequest) Flush(_ context.Context) error {
+
+	_, end := giteaCommentLineRange(c)
+	return end
+}
+
+func giteaCommentLineRange(c *reviewdog.Comment) (start int, end int) {
+	var rdfRange *rdf.Range
+
+	// Prefer first suggestion line range to diagnostic location if available so
+	// that reviewdog can post code suggestion as well when the line ranges are
+	// different between the diagnostic location and its suggestion.
+	if c.Result.FirstSuggestionInDiffContext && len(c.Result.Diagnostic.GetSuggestions()) > 0 {
+		rdfRange = c.Result.Diagnostic.GetSuggestions()[0].GetRange()
+	} else {
+		rdfRange = c.Result.Diagnostic.GetLocation().GetRange()
+	}
+
+	startLine := rdfRange.GetStart().GetLine()
+	endLine := rdfRange.GetEnd().GetLine()
+	if endLine == 0 {
+		endLine = startLine
+	}
+	return int(startLine), int(endLine)
+}
+
+func (g *PullRequest) remainingCommentsSummary(remaining []*reviewdog.Comment, baseURL string, gitRootPath string) string {
+	if len(remaining) == 0 {
+		return ""
+	}
+	perTool := make(map[string][]*reviewdog.Comment)
+	for _, c := range remaining {
+		perTool[c.ToolName] = append(perTool[c.ToolName], c)
+	}
+	var sb strings.Builder
+	sb.WriteString("Remaining comments which cannot be posted as a review comment to avoid spamming Pull Request\n")
+	sb.WriteString("\n")
+	for tool, comments := range perTool {
+		sb.WriteString("<details>\n")
+		sb.WriteString(fmt.Sprintf("<summary>%s</summary>\n", tool))
+		sb.WriteString("\n")
+		for _, c := range comments {
+			sb.WriteString("<hr>")
+			sb.WriteString("\n")
+			sb.WriteString("\n")
+			sb.WriteString(commentutil.MarkdownComment(c))
+			sb.WriteString("\n")
+			sb.WriteString("\n")
+			sb.WriteString(giteaCodeSnippetURL(baseURL, gitRootPath, c.Result.Diagnostic.GetLocation()))
+			sb.WriteString("\n")
+			sb.WriteString("\n")
+		}
+		sb.WriteString("</details>\n")
+	}
+	return sb.String()
+}
+
 // setPostedComment get posted comments from Gitea.
 func (g *PullRequest) setPostedComment() error {
 	g.postedcs = make(commentutil.PostedComments)
-
+	g.outdatedComments = make(map[string]*gitea.PullReviewComment)
+	g.prCommentWithReply = make(map[int64]bool)
 	cs, err := g.comment()
 	if err != nil {
 		return err
 	}
 
+	commentThreads := make(map[string]int64, len(cs)) // commit/path:line
 	for _, c := range cs {
-		if c.LineNum == 0 || c.Path == "" || c.Body == "" {
-			continue
+		commentKey := fmt.Sprintf("%s/%s:%d", c.CommitID, c.Path, c.LineNum)
+		replyID, ok := commentThreads[commentKey]
+		if !ok {
+			commentThreads[commentKey] = c.ID
+		} else {
+			g.prCommentWithReply[replyID] = true
+		}
+
+		if meta := extractMetaComment(c.Body); meta != nil {
+			g.postedcs.AddPostedComment(c.Path, int(c.LineNum), meta.GetFingerprint())
+			if meta.SourceName == g.toolName {
+				g.outdatedComments[meta.GetFingerprint()] = c // Remove non-outdated comment later.
+			}
 		}
-		g.postedcs.AddPostedComment(c.Path, int(c.LineNum), c.Body)
 	}
+	return nil
+}
 
+func extractMetaComment(body string) *metacomment.MetaComment {
```

Comment:
Can you put this func under service/serviceutil/meta.go and re-use the same implementation between github and gitea reporters instead of copying functions?

Reply:
Sure, will do that